django/advanced-guides/staging-environments-with-github-actions.html.md

```diff
@@ -0,0 +1,336 @@
+`https://pr-9-felixxm-fly-pr-preview-example.fly.dev/`
+</div>
+
+(for PR number 9 in the repository called `pr-preview-example`).
+
+## Using Postgres cluster
+
+Using a dedicated staging database is a good practice for test environments. This gives
+us more control and appropriate separation from the production environment which
+eliminates a potential data leak vector.
+
+If you don't have a [Postgres cluster](https://fly.io/docs/postgres/) specifically for
+testing purposes you can create one with [`fly postgres create`](https://fly.io/docs/flyctl/postgres-create/):
+
+```cmd
+fly postgres create --name pg-fly-pr-staging-preview
+```
+
+Once created, we can specify it in our action (`jobs → preview-app → steps → with`)
+using the `postgres` input:
+
+```yaml
+# .github/workflows/fly_pr_preview.yml
+
+name: Start preview app
+
+on:
+  pull_request:
+    types: [labeled, synchronize, opened, reopened, closed]
+
+concurrency:
+  group: ${{ github.workflow }}-pr-${{ github.event.number }}
+  cancel-in-progress: true
+
+permissions:
+  contents: read
+
+env:
+  FLY_API_TOKEN: ${{ secrets.FLY_API_TOKEN }}
+
+jobs:
+  preview-app:
+    if: contains(github.event.pull_request.labels.*.name, 'PR preview app')
+    runs-on: ubuntu-latest
+    name: Preview app
+    environment:
+      name: pr-${{ github.event.number }}
+      url: ${{ steps.deploy.outputs.url }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Deploy preview app
+        uses: superfly/fly-pr-review-apps@1.2.0
+        id: deploy
+        with:
+          postgres: pg-fly-pr-staging-preview  # ← Added
+          region: waw
+          org: personal
+```
+
+With that in place, our staging Postgres cluster will be automatically attached to the
+test app, which will make a `DATABASE_URL` environment variable available in the test
+VM.
+
+## Additional release steps
+
+For performing additional release steps when deploying our staging environment, we can
+use a custom [Fly.io TOML configuration file](https://fly.io/docs/reference/configuration/)
+dedicated for staging with a release script
+[to run before a deployment](https://fly.io/docs/reference/configuration/#run-one-off-commands-before-releasing-a-deployment).
+First, make a copy of an existing configuration:
+
+```cmd
+mkdir staging
+```
+```cmd
+cp fly.toml staging/fly_staging.toml
+```
+
+Next, create a script (`staging/post_deploy.sh`) to prepare our staging database:
+
+```bash
+# staging/post_deploy.sh
+#!/usr/bin/env bash
+
+# Migrate database.
+python /code/manage.py migrate
+# Load fixtures with test data.
+python /code/manage.py loaddata /code/staging/test_groups.json
+```
+
+Finally, we need to add `release_command` to the TOML configuration calling our script:
+
+```toml
+# staging/fly_staging.toml
+
+console_command = "/code/manage.py shell"
+
+[build]
+
+[env]
+  PORT = "8000"
+
+[deploy]
+  release_command = "sh ./staging/post_deploy.sh" # ← Added.
+
+...
```

Comment:
A typo here

Reply:
What kind of typo? :thinking: I wanted to leave ellipsis because that's not the complete TOML file.

Reply:
Aha! Sorry, I thought it was ```. All good!